### ISSUE TITLE
spec: Consensus Write-Ahead Log (WAL)

### ISSUE THREAD

**Issue by cason** [circlefin/malachite#469]
In order to support the crash-recovery failure model (#578), the consensus implementation should persist all relevant events that have lead it to its current state. When recovering from a crash, the implementation is initialized from its initial state of the latest active height `H`, then has to replay all the information persisted before it has crashed.

The valid events processed by the consensus implementation are therefore typically persisted in a Write-Ahead Log (WAL), an append-only log that was originally conceived to ensure atomicity of transactions in databases.

## Definition of Done

- [ ] Write an ADR discussing the problem and defining:
   - Which events should be persisted to the WAL
   - The strategy for writing data to the WAL (e.g., synchronously versus asynchronously)
   - The procedure for replaying the content of the WAL when a node is (re)started
- [ ] Check if the implementation aligns with the design

**Comment by cason** [comment]
Besides of better documenting the solution, can we consider it solved?

**Comment by cason** [comment]
Hey folks, do you remember what was the designed procedure to handle `getValue()` calls in a recovering process? How is this handled in the implementation?

In am particular interested in the case where a recovering process was the proposer of multiple rounds of a height before crashing, therefore potentially has provided multiple distinct values to consensus as a response/return to `getValue()`.

Also in this case, broadcasting again values proposed in rounds that have failed is not really ideal, isn't it?

**Comment by josef-widder** [comment]
I guess we talked about having another persistent blockstore, where we put the proposals seen/generated for the current height. In contrast to the real blockstore, the proposal store can be deleted in the next height.

**Comment by cason** [comment]
> I guess we talked about having another persistent blockstore, where we put the proposals seen/generated for the current height. In contrast to the real blockstore, the proposal store can be deleted in the next height.

Yes, we have this as part of https://github.com/informalsystems/malachite/issues/579. But how we handle the case where `getValue()` is called multiple times upon recovery? How the "application" knows which one to return? Does it re-broadcast every "potential" proposed value again upon recovery?

@romac, @ancazamfir, how this work in the implementation?

**Comment by cason** [comment]
https://github.com/informalsystems/malachite/issues/579#issuecomment-2612044335

If this is the case, I would say that half of my question is solved. Namely, the "application" is able to know which proposed value to return in case of multiple rounds.

It remains open whether all the potential multiple proposal values are re-broadcast/streamed.
